Allowed non-const args for randDistrib

pinsvin00 opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added possibility to add non constant arguments for randUniform, randNormal, randLogNormal, randExponential, randChiSquared, randStudentT, randFisherF, randBernoulli, randBinomial, randNegativeBinomial, randPoisson
### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [ ] <!---ci_exclude_fast--> Exclude: Fast test
- [ ] <!---ci_exclude_asan--> Exclude: All with ASAN
- [ ] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [ ] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

pinsvin00 committed:
Allowed non-const args for randDistrib

pinsvin00 commented:
Will add tests soon, don't have any idea to write them atm

pinsvin00 commented on src/Functions/randDistribution.cpp:
```diff
@@ -226,11 +225,20 @@ class FunctionRandomDistribution : public IFunction
                             parameter_number, arguments.size());
 
         const IColumn * col = arguments[parameter_number].column.get();
+        ResultType parameter;
```
Shouldn't be ResultType, if argument will be Int64, and result Float64, it will cause invalid execution

vdimir commented:
> Will add tests soon, don't have any idea to write them atm

You may add cases to sql test https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/02462_distributions.sql and update `.referece` accrodingly

more info https://clickhouse.com/docs/en/development/tests

pinsvin00 committed:
Added better logging for invalid parameters, added test case

pinsvin00 committed:
Merge branch 'master' into add-non-constant-args-randDistrib

pinsvin00 commented:
Ok, updated the reference and the test. Should be ok, although I am worried a little bit about randFisherF(), randStudentT() and randChiSquared(), when we run randFisherF(1 + (rand() %1)), or any non constant expression, the function generate() is being called a couple of times, and in the first calls, the argument colmuns aren't interpreted yet, and they hold value 0, which casues generate() func to throw an expression. When I removed the code responsible for checking degree of freedom, it works correctly. Should I be concerned of this at this point? Or it could be fixed by some separate ticket?
Cheers

```
    static void generate(Float64 d1, Float64 d2, ColumnFloat64::Container & container)
    {
        if (d1 <= 0 || d2 <= 0) // this check is causing an error
            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Argument (degrees of freedom) of function {} should be greater than zero", getName());

        auto distribution = std::fisher_f_distribution<>(d1, d2);
        for (auto & elem : container)
            elem = distribution(thread_local_rng);
    }
```

pinsvin00 commented:
I've checked it also for LEFT() function, some of the first calls are having 0 param.
![image](https://github.com/user-attachments/assets/9a5e7ede-832e-4c25-b8e1-1c4126d30e10)

pinsvin00 committed:
Fixed issues with degrees of freedom function.

pinsvin00 commented:
Ok, managed to find a fix for it

robot-clickhouse commented:
<!-- automatic status comment for PR #68982 from pinsvin00/ClickHouse:add-non-constant-args-randDistrib -->
*This is an automated comment for commit e81516bdc68f29fc214ae51973d43b299637b7b8 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68982/e81516bdc68f29fc214ae51973d43b299637b7b8/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

pinsvin00 committed:
Fixed style check

pinsvin00 committed:
Added missing newline in reference 02462_distributions.reference

pinsvin00 committed:
Updated minimal value for distributions tests

pinsvin00 committed:
Merge branch 'master' into add-non-constant-args-randDistrib

nikitamikhaylov commented on src/Functions/randDistribution.cpp:
```diff
@@ -226,11 +224,27 @@ class FunctionRandomDistribution : public IFunction
+        }
+        else
+        {
+            std::string expectedType;
+            if (std::is_same_v<UInt64, ResultType>)
+                expectedType = "UInt64";
+            else if (std::is_same_v<Float64, ResultType>)
+                expectedType = "Float64";
+
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Parameter number {} of function {} is expected to be {} but is {}",
+                parameter_number, getName(), expectedType, col->getName());
```
It might be more convenient to use:
```
Types::Which(TypeToEnum<std::decay_t<T>>::value)->getTypeName();
```

nikitamikhaylov commented on src/Functions/randDistribution.cpp:
```diff
@@ -93,7 +92,7 @@ struct ChiSquaredDistribution
     static void generate(Float64 degree_of_freedom, ColumnFloat64::Container & container)
     {
-        if (degree_of_freedom <= 0)
+        if (!container.empty() && degree_of_freedom <= 0)
```
Why? The `degree_of_freedom` is parameter of distribution. We should check unconditionally.

pinsvin00 replied:
I've checked it and when we run for example randFisherF(1 + (rand() % 1), 2), some of the first calls of this generate)_, contain ColumnFloat64::Container which is empty. I've just added this check so we exclude this check, when 1 + (rand() % 1) isn't resolved yet (container is empty, and degree_of_freedom = 0).

nikitamikhaylov committed:
Beautify

nikitamikhaylov commented:
The `AST fuzzer (msan)` failure is related, let's try to identify an issue.

pinsvin00 committed:
Merge branch 'master' into add-non-constant-args-randDistrib

pinsvin00 committed:
Merge branch 'master' into add-non-constant-args-randDistrib

pinsvin00 committed:
Attempt to fix msan, complaint

nikitamikhaylov commented:
`AST fuzzer (msan)` is still failing.

alexey-milovidov commented:
@pinsvin00, let's check the failures.

pinsvin00 committed:
Second attempt to fix MSan complaint

pinsvin00 committed:
Fixed style check

alexey-milovidov commented:
It is still failed.
